### ISSUE TITLE
Echo handler not working with HTTP/2 server

### ISSUE THREAD

**Issue by bclozel** [reactor/reactor-netty#395]
I've got a simple echo reactor handler that works fine with HTTP1.1 but not with HTTP/2.
Is my echo handler wrong? Or is there something specific with the handling of HTTP/2 streams that doesn't work with this?

Given the following code snippet:

```java
BiFunction<HttpServerRequest, HttpServerResponse, Mono<Void>> sampleHandler =
		(req, res) -> res.sendString(Flux.just("Spring Boot")).then();

SelfSignedCertificate cert = new SelfSignedCertificate();
DisposableServer server = HttpServer.create()
		.secure(spec -> {
			spec.sslContext(SslContextBuilder.forServer(cert.certificate(), cert.privateKey()))
					.defaultConfiguration(SslProvider.DefaultConfigurationType.H2);
		})
		.protocol(HttpProtocol.H2)
		.handle(sampleHandler)
		.port(8080)
		.bindNow();
```

The HTTP/2 server starts fine and the following curl request works just fine as well:

```
curl --http2 --insecure -vv https://localhost:8080/test -H "Content-Type: text/plain" -d "Spring Boot2"
```

Here are the server logs:

```
16:14:41.949 [main] DEBUG r.netty.resources.DefaultLoopEpoll - Default Epoll support : false
16:14:41.955 [main] DEBUG r.netty.resources.DefaultLoopKQueue - Default KQueue support : false
16:14:42.061 [reactor-http-nio-1] DEBUG reactor.netty.tcp.TcpServer - [id: 0xf042b008, L:/0:0:0:0:0:0:0:1:8080] Bound new server
16:14:45.816 [reactor-http-nio-2] DEBUG reactor.netty.tcp.SslProvider - [id: 0x54f0b3ed, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51644] SSL enabled using engine OpenSslEngine
16:14:45.879 [reactor-http-nio-2] DEBUG r.netty.channel.BootstrapHandlers - [id: 0x54f0b3ed, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51644] Initialized pipeline DefaultChannelPipeline{(reactor.left.sslHandler = io.netty.handler.ssl.SslHandler), (reactor.left.sslReader = reactor.netty.tcp.SslProvider$SslReadHandler), (BootstrapHandlers$BootstrapInitializerHandler#0 = reactor.netty.channel.BootstrapHandlers$BootstrapInitializerHandler), (reactor.left.httpCodec = io.netty.handler.codec.http2.Http2MultiplexCodec), (reactor.right.reactiveBridge = reactor.netty.channel.ChannelOperationsHandler)}
16:14:45.899 [reactor-http-nio-2] DEBUG r.n.channel.ChannelOperationsHandler - [id: 0x54f0b3ed, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51644] No ChannelOperation attached. Dropping: DefaultHttp2SettingsFrame(settings={ENABLE_PUSH=0, MAX_CONCURRENT_STREAMS=100, INITIAL_WINDOW_SIZE=1073741824})
16:14:45.921 [reactor-http-nio-2] DEBUG r.n.http.server.HttpServerOperations - [id: 0x54f0b3ed, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51644](H2 - 1) New http2 connection, requesting read
16:14:45.928 [reactor-http-nio-2] DEBUG reactor.netty.http.server.HttpServer - [id: 0x54f0b3ed, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51644](H2 - 1) Initialized HTTP/2 pipeline Http2MultiplexCodec$DefaultHttp2StreamChannel$1{(HttpServerBind$Http2StreamInitializer#0 = reactor.netty.http.server.HttpServerBind$Http2StreamInitializer), (Http2StreamBridgeHandler#0 = reactor.netty.http.server.Http2StreamBridgeHandler), (Http2StreamFrameToHttpObjectCodec#0 = io.netty.handler.codec.http2.Http2StreamFrameToHttpObjectCodec), (reactor.right.reactiveBridge = reactor.netty.channel.ChannelOperationsHandler)}
16:14:45.983 [reactor-http-nio-2] DEBUG reactor.netty.http.server.HttpServer - [id: 0x54f0b3ed, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51644](H2 - 1) Handler is being applied: Scratch$$Lambda$2/6566818@1262a6a0
16:14:45.994 [reactor-http-nio-2] DEBUG r.n.channel.ChannelOperationsHandler - [id: 0x54f0b3ed, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51644](H2 - 1) Writing object DefaultHttpResponse(decodeResult: success, version: HTTP/1.1)
HTTP/1.1 200 OK
16:14:46.000 [reactor-http-nio-2] DEBUG r.n.channel.ChannelOperationsHandler - [id: 0x54f0b3ed, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51644](H2 - 1) Writing object MonoCallable
16:14:46.003 [reactor-http-nio-2] DEBUG r.n.http.server.HttpServerOperations - [id: 0x54f0b3ed, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51644](H2 - 1) Last HTTP response frame
16:14:46.004 [reactor-http-nio-2] DEBUG r.n.channel.ChannelOperationsHandler - [id: 0x54f0b3ed, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51644](H2 - 1) Writing object EmptyLastHttpContent
16:14:46.007 [reactor-http-nio-2] DEBUG r.n.channel.ChannelOperationsHandler - [id: 0x54f0b3ed, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51644](H2 - 1) Write state change false
```

Now with an echo reactor handler, like this:

```java
BiFunction<HttpServerRequest, HttpServerResponse, Mono<Void>> echoHandler =
		(req, res) -> res.send(req.receive().map(ByteBuf::retain)).then();

SelfSignedCertificate cert = new SelfSignedCertificate();
DisposableServer server = HttpServer.create()
		.secure(spec -> {
			spec.sslContext(SslContextBuilder.forServer(cert.certificate(), cert.privateKey()))
					.defaultConfiguration(SslProvider.DefaultConfigurationType.H2);
		})
		.protocol(HttpProtocol.H2)
		.handle(echoHandler)
		.port(8080)
		.bindNow();
```

the same curl request just hangs and never receives the body:

```
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x7fa6fc004800)
> POST /test HTTP/2
> Host: localhost:8080
> User-Agent: curl/7.54.0
> Accept: */*
> Content-Type: text/plain
> Content-Length: 12
>
* Connection state changed (MAX_CONCURRENT_STREAMS updated)!
* We are completely uploaded and fine
< HTTP/2 200
```

Here are the server logs:

```
16:16:22.989 [main] DEBUG r.netty.resources.DefaultLoopEpoll - Default Epoll support : false
16:16:22.996 [main] DEBUG r.netty.resources.DefaultLoopKQueue - Default KQueue support : false
16:16:23.096 [reactor-http-nio-1] DEBUG reactor.netty.tcp.TcpServer - [id: 0x365ff518, L:/0:0:0:0:0:0:0:1:8080] Bound new server
16:16:26.758 [reactor-http-nio-2] DEBUG reactor.netty.tcp.SslProvider - [id: 0xdac8713b, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51657] SSL enabled using engine OpenSslEngine
16:16:26.815 [reactor-http-nio-2] DEBUG r.netty.channel.BootstrapHandlers - [id: 0xdac8713b, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51657] Initialized pipeline DefaultChannelPipeline{(reactor.left.sslHandler = io.netty.handler.ssl.SslHandler), (reactor.left.sslReader = reactor.netty.tcp.SslProvider$SslReadHandler), (BootstrapHandlers$BootstrapInitializerHandler#0 = reactor.netty.channel.BootstrapHandlers$BootstrapInitializerHandler), (reactor.left.httpCodec = io.netty.handler.codec.http2.Http2MultiplexCodec), (reactor.right.reactiveBridge = reactor.netty.channel.ChannelOperationsHandler)}
16:16:26.835 [reactor-http-nio-2] DEBUG r.n.channel.ChannelOperationsHandler - [id: 0xdac8713b, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51657] No ChannelOperation attached. Dropping: DefaultHttp2SettingsFrame(settings={ENABLE_PUSH=0, MAX_CONCURRENT_STREAMS=100, INITIAL_WINDOW_SIZE=1073741824})
16:16:26.858 [reactor-http-nio-2] DEBUG r.n.http.server.HttpServerOperations - [id: 0xdac8713b, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51657](H2 - 1) New http2 connection, requesting read
16:16:26.864 [reactor-http-nio-2] DEBUG reactor.netty.http.server.HttpServer - [id: 0xdac8713b, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51657](H2 - 1) Initialized HTTP/2 pipeline Http2MultiplexCodec$DefaultHttp2StreamChannel$1{(HttpServerBind$Http2StreamInitializer#0 = reactor.netty.http.server.HttpServerBind$Http2StreamInitializer), (Http2StreamBridgeHandler#0 = reactor.netty.http.server.Http2StreamBridgeHandler), (Http2StreamFrameToHttpObjectCodec#0 = io.netty.handler.codec.http2.Http2StreamFrameToHttpObjectCodec), (reactor.right.reactiveBridge = reactor.netty.channel.ChannelOperationsHandler)}
16:16:26.927 [reactor-http-nio-2] DEBUG reactor.netty.http.server.HttpServer - [id: 0xdac8713b, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51657](H2 - 1) Handler is being applied: Scratch$$Lambda$1/930990596@397a4bd
16:16:26.933 [reactor-http-nio-2] DEBUG r.n.channel.ChannelOperationsHandler - [id: 0xdac8713b, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51657](H2 - 1) Writing object DefaultHttpResponse(decodeResult: success, version: HTTP/1.1)
HTTP/1.1 200 OK
16:16:26.940 [reactor-http-nio-2] DEBUG r.n.channel.ChannelOperationsHandler - [id: 0xdac8713b, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51657](H2 - 1) Writing object 
16:16:26.941 [reactor-http-nio-2] DEBUG reactor.netty.channel.FluxReceive - [id: 0xdac8713b, L:/0:0:0:0:0:0:0:1:8080 - R:/0:0:0:0:0:0:0:1:51657](H2 - 1) Subscribing inbound receiver [pending: 0, cancelled:false, inboundDone: false]
```

Note: Even with HTTP/2, the server logs are still mentioning `HTTP/1.1`

**Comment by violetagg** [comment]
temporary fix provided for 0.8.0.M1

**Comment by violetagg** [comment]
The temporary fix was reverted. Netty is updated to version 4.1.28.Final where the fix exists.
